init-container/proto.h

```diff
@@ -0,0 +1,131 @@
+#ifndef _PROTO_H
+#define _PROTO_H
+
+#include <stdint.h>
+
+#pragma pack(push, 1)
+
+/*
+ * Host -> Guest
+ *
+ * - u64 message ID (non zero),
+ * - 1 byte type,
+ * - stream of sub-messages ended with `SUB_MSG_END`; each sub-message
+ *   consists of 1 byte subtype subtype-specific body (described near each
+ *   subtype below) of types:
+ * BYTES - unsigned 64bit length, followed by exactly that many bytes;
+ * ARRAY - unsigned 64bit number of BYTES objects, that follow immediately
+ *         (encoded as described above);
+ * u64   - unsigned 64bit number,
+ * u32   - unsigned 32bit number.
+ *
+ * All numbers are encoded in little-endian format.
+ */
+
+/*
+ * Guest -> Host
+ *
+ * Guest sends two types of messages - response and asynchronous notification:
+ * - u64: message ID matching the request for response, 0 for notification,
+ * - 1 byte type, followed by type-specific body.
+ */
+
+typedef uint64_t msg_id_t;
+
+struct msg_hdr {
+    msg_id_t msg_id;
+    uint8_t type;
+};
+
+enum MSG_TYPE {
+    MSG_QUIT = 1,
+    /* Returns ID of spawned process. */
+    MSG_RUN_PROCESS,
+    MSG_KILL_PROCESS,
+    MSG_MOUNT_VOLUME,
+    MSG_UPLOAD_FILE,
+    /* Returns chunk of process' output as BYTES. */
+    MSG_QUERY_OUTPUT,
+    MSG_PUT_INPUT,
+    MSG_SYNC_FS,
+};
+
+enum MSG_SUB_TYPE {
```

Comment:
I think that subtypes should be separate for each type. Something like

```c
enum MSG_SUB_TYPE_RUN_PROCESS {
    /* End of sub-messages. */
    SUB_MSG_END = 0,

    /* Binary path. (BYTES) */
    SUB_MSG_RUN_PROCESS_BIN,
    [...]
};

enum MSG_SUB_TYPE_KILL_PROCESS {
    /* End of sub-messages. */
    SUB_MSG_END = 0,

    /* ID of process. (u64) */
    SUB_MSG_KILL_PROCESS_ID,
};

[...]
```

Reply:
@boryspoplawski Maybe this spec is large enough to use: https://github.com/nanopb/nanopb ?

Reply:
As discussed off GH, we keep current version for now and probably move to protobuf later on.